magento/magento2
# Don't disable FPC for maintenance, instead send "no cache" headers

### Description (*)
This Pull Request removes the need to have a writable env.php while switching the maintenance mode. Previously Magento tried to disable the FPC when switching into maintenance mode. This removes that and instead, to maintain functionality, sets the "no-cache" headers for all frontend requests in maintenance mode (``Cache-Control`` etc.). The effect is, that pages are not cached while in maintenance mode.

- Removes the observer for disabling and re-enabling the FPC during
  maintenande mode switch

- Adds a new observer, that adds the "no cache" headers for all HTTP
  requests during maintenance mode

### Fixed Issues (if relevant)
1. magento/magento2#24229: Unable to enable maintenance mode when env.php is read only

### Manual testing scenarios (*)
With Builtin FPC:
1. Keep maintenance mode disabled, access a cacheable Magento page
    - Page should be cached and delivered by FPC on next request
2. Make the app/etc/env.php file read-only
3. Enable maintenance mode
    - This should now not fail, previously it complained about the env.php not being writable
4. Add the local IP to the maintenance whitelist
5. Flush the FPC
6. Access a cacheable Magento page
    - Page should not be cached
7. Disable maintenance again and repeat 1.

Repeat tests for Varnish as FPC.

### Contribution checklist (*)
